cloud/aws/runtime/queue/sqs.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2021 Nitric Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
+	"github.com/aws/smithy-go"
+	structpb "github.com/golang/protobuf/ptypes/struct"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-sdk-go-v2/otelaws"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/nitrictech/nitric/cloud/aws/ifaces/sqsiface"
+	"github.com/nitrictech/nitric/cloud/aws/runtime/env"
+	"github.com/nitrictech/nitric/cloud/aws/runtime/resource"
+	grpc_errors "github.com/nitrictech/nitric/core/pkg/grpc/errors"
+
+	queuepb "github.com/nitrictech/nitric/core/pkg/proto/queues/v1"
+)
+
+const (
+	// ErrCodeNoSuchTagSet - AWS API neglects to include a constant for this error code.
+	ErrCodeNoSuchTagSet = "NoSuchTagSet"
+	ErrCodeAccessDenied = "AccessDenied"
+)
+
+type SQSQueueService struct {
+	provider resource.AwsResourceProvider
+	client   sqsiface.SQSAPI
+}
+
+var _ queuepb.QueuesServer = &SQSQueueService{}
+
+// Get the URL for a given queue name
+func (s *SQSQueueService) getUrlForQueueName(ctx context.Context, queue string) (*string, error) {
+	queues, err := s.provider.GetResources(ctx, resource.AwsResource_Queue)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving queue list")
+	}
+
+	queueArn, ok := queues[queue]
+
+	if !ok {
+		return nil, fmt.Errorf("queue %s does not exist", queue)
+	}
+
+	arnParts := strings.Split(queueArn.ARN, ":")
+	accountId := arnParts[4]
+	queueName := arnParts[5]
+
+	out, err := s.client.GetQueueUrl(ctx, &sqs.GetQueueUrlInput{
+		QueueName:              aws.String(queueName),
+		QueueOwnerAWSAccountId: aws.String(accountId),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("encountered an error retrieving the queue list: %w", err)
+	}
+
+	return out.QueueUrl, nil
+}
+
+func isSQSAccessDeniedErr(err error) bool {
+	var opErr *smithy.OperationError
+	if errors.As(err, &opErr) {
+		return opErr.Service() == "SQS" && strings.Contains(opErr.Unwrap().Error(), "AccessDenied")
+	}
+	return false
+}
+
+func (s *SQSQueueService) Send(ctx context.Context, req *queuepb.QueueSendRequestBatch) (*queuepb.QueueSendResponse, error) {
```

Comment:
Should we update the proto to only have a 'QueueSend' req/res, which can include multiple items and remove the batch operation names (or vice versa)?